blockchain/utxocache.go

```diff
@@ -0,0 +1,1085 @@
+//
+// This method should be called with the state lock held.
+func (s *utxoCache) fetchAndCacheEntries(ops []wire.OutPoint) ([]*UtxoEntry, error) {
+	entries := make([]*UtxoEntry, len(ops))
+	err := s.db.View(func(dbTx database.Tx) error {
+		utxoBucket := dbTx.Metadata().Bucket(utxoSetBucketName)
+
+		for i, op := range ops {
+			entry, err := dbFetchUtxoEntry(dbTx, utxoBucket, op)
+			if err != nil {
+				return err
+			}
+
+			entries[i] = entry
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Add each of the entries to the UTXO cache and update their memory
+	// usage.
+	//
+	// NOTE: When the fetched entry is nil, it is still added to the cache
+	// as a miss; this prevents future lookups to perform the same database
+	// fetch.
+	for i, entry := range entries {
+		s.totalEntryMemory += entry.memoryUsage()
+		s.cachedEntries.put(ops[i], entry, s.totalEntryMemory)
+	}
+
+	return entries, nil
+}
+
+// getEntry returns the UTXO entry for the given outpoint.  It returns nil if
+// there is no entry for the outpoint in the UTXO state.
+//
+// This method is part of the utxoView interface.
+// This method should be called with the state lock held.
+// The returned entry is NOT safe for concurrent access.
+func (s *utxoCache) getEntry(outpoint wire.OutPoint) (*UtxoEntry, error) {
+	// First, we'll check out in-memory cache to see if we already have the
+	// entry.
+	entry, found := s.cachedEntries.get(outpoint)
+	if found {
+		return entry, nil
+	}
+
+	// If not, then we'll fetch it from the databse.
+	entries, err := s.fetchAndCacheEntries([]wire.OutPoint{outpoint})
+	if err != nil {
+		return nil, err
+	}
+
+	return entries[0], nil
+}
+
+// getEntry returns the UTXO entry for the given outpoint.  It returns nil if
+// there is no entry for the outpoint in the UTXO state.
+//
+// This method is part of the utxoView interface.
+// This method should be called with the state lock held.
+// The returned entry is NOT safe for concurrent access.
+func (s *utxoCache) getEntries(outpoints []wire.OutPoint) ([]*UtxoEntry, error) {
+	entries := make([]*UtxoEntry, len(outpoints))
+	missingOps := make([]wire.OutPoint, 0, len(outpoints))
+	missingOpsIdx := make([]int, 0, len(outpoints))
+	for i, op := range outpoints {
+		if entry, ok := s.cachedEntries.get(op); ok {
+			entries[i] = entry
+			continue
+		}
+
+		missingOpsIdx = append(missingOpsIdx, i)
+		missingOps = append(missingOps, op)
+	}
+
+	dbEntries, err := s.fetchAndCacheEntries(missingOps)
+	if err != nil {
+		return nil, err
+	}
+
+	for i, idx := range missingOpsIdx {
+		entries[idx] = dbEntries[i]
+	}
+
+	return entries, nil
+}
+
+// FetchEntry returns the UTXO entry for the given outpoint.  It returns nil if
+// there is no entry for the outpoint in the UTXO state.
+//
+// This method is safe for concurrent access.
+func (s *utxoCache) FetchEntry(outpoint wire.OutPoint) (*UtxoEntry, error) {
+	s.mtx.Lock()
+	entries, err := s.getEntries([]wire.OutPoint{outpoint})
```

Comment:
suggestion: Use the `getEntry` method, or remove the `getEntry` logic and use `getEntries` everywhere.

Reply:
Ah read this comment after I posted https://github.com/btcsuite/btcd/pull/1955#discussion_r1165018164

Will do so.